crates/hyperqueue/src/dashboard/ui/screens/job/job_tasks_chart.rs

```diff
@@ -0,0 +1,155 @@
+        if let Some(job_id) = self.job_id {
+            let (num_running_tasks, num_failed_tasks, num_finished_tasks) =
+                get_task_counts_for_job(job_id, data, SystemTime::now());
+            self.chart.set_chart_name(
+                format!("Running: {num_running_tasks}, Finished: {num_finished_tasks}, Failed: {num_failed_tasks}").as_str(),
+            );
+        } else {
+            self.chart
+                .set_chart_name("Please select a Job to show Task counts");
+        }
+        self.chart.update(data);
+    }
+
+    pub fn draw(&mut self, rect: Rect, frame: &mut DashboardFrame) {
+        self.chart.draw(rect, frame);
+    }
+}
+
+struct TaskCountsPlotter(JobId);
+
+impl ChartPlotter for TaskCountsPlotter {
+    fn get_charts(&self) -> Map<String, PlotStyle> {
+        let mut plots_and_styles: Map<String, PlotStyle> = Default::default();
+
+        let running_points_style = PlotStyle {
+            color: Color::Yellow,
+            marker: Marker::Braille,
+        };
+        let finished_points_style = PlotStyle {
+            color: Color::Green,
+            marker: Marker::Braille,
+        };
+        let failed_points_style = PlotStyle {
+            color: Color::Red,
+            marker: Marker::Braille,
+        };
+
+        plots_and_styles.insert(RUNNING_TAG.to_string(), running_points_style);
+        plots_and_styles.insert(FINISHED_TAG.to_string(), finished_points_style);
+        plots_and_styles.insert(FAILED_TAG.to_string(), failed_points_style);
+
+        plots_and_styles
+    }
+
+    /// Fetches the number of Running/Finished/Failed tasks at a given time and makes it available to the chart.
+    fn fetch_data_points(
+        &self,
+        data: &DashboardData,
+        at_time: SystemTime,
+    ) -> Map<String, (f64, f64)> {
+        let mut data_points: Map<String, (f64, f64)> = Default::default();
+
+        let (num_running_tasks, num_failed_tasks, num_finished_tasks) =
+            get_task_counts_for_job(self.0, data, at_time);
+        if num_running_tasks != 0 {
+            data_points.insert(
+                RUNNING_TAG.to_string(),
+                (get_time_as_secs(at_time), num_running_tasks as f64),
+            );
+        }
+        if num_finished_tasks != 0 {
+            data_points.insert(
+                FINISHED_TAG.to_string(),
+                (get_time_as_secs(at_time), num_finished_tasks as f64),
+            );
+        }
+        if num_failed_tasks != 0 {
+            data_points.insert(
+                FAILED_TAG.to_string(),
+                (get_time_as_secs(at_time), num_failed_tasks as f64),
+            );
+        }
+
+        data_points
+    }
+}
+
+fn get_task_counts_for_job(
+    job_id: JobId,
+    data: &DashboardData,
+    at_time: SystemTime,
+) -> (usize, usize, usize) {
+    let num_running_tasks =
+        fetch_job_task_count_with_status(job_id, data, at_time, DashboardTaskState::Running);
```

Comment:
This seems quite ineffective, since the query is invoked three times. Let's just invoke it once, iterate the items and split them by state.